### DISCUSSION TITLE
Setting Modulator values with another Modulator

### DISCUSSION THREAD

**Discussion by DRBAUS** [RomanKubiak/ctrlr#125] [General]
I am still working on a panel and have the following issue:

There are 8 layers in the panel. Each Modulator in each panel has it's own script and that is all working. 

What I would like is to create a 9th layer which represents an EDIT ALL layers. So my idea is that I can create a script for a Modulator in Layer 9, let's call that Modulator ENV1 Attack, which in turn when set to a value of for example 99 sets all the Modulators that represent ENV1 in the other 8 Layers to 99. Those Modulators in Layer 1-8 will then fire their script and send their sysex to the synth.

This idea is my 2nd option. My first option was to receive a Parameter Block from the synth and have that update the Modulator Values in Ctrlr but unfortunately that won't work. The sysex of this unit is too complicated for that.

What I have going now for testing is the following and it does not work:

WSALL_ENV1_TIME1 = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

value = panel:getModulatorByName("WSALL_ENV1_TIME_1"):getModulatorValue() --this is the one Modulator to edit the others --

if value==0 then
panel:getModulatorByName("WS1_ENV1_TIME_1"):setValue(0,true,true)

elseif value==99 then
panel:getModulatorByName("WS1_ENV1_TIME_1"):setValue(99,true,true)

What part am I missing?

As usual thanks in advance for your input.
end
end

**Comment by Ted-juh** [comment]
setValue(99,true) or setModulatorValue(99, true, true)

If you need that many if-statements I would use a for loop. 

for value = 0,127 do
panel:getModulatorByName(): setModulatorValue (value,true, true)
end

There is a post on the Ctrlr forums by Goodweather (and a newer post from Dnaldoog that points to that post) about the booleans and setValue/ setModulatorValue. Some do misfire or simply don't work.

**Comment by dobo365** [comment]
Hi DRBAUS,
 different things here...
- setValue(value, true) is the way to go
- in this when modulator value change method (I'm calling it an OnChange), you don't need to retrieve value as it is directly available from the arguments of the method
So, you first statement value = panel:getModulatorByName("WSALL_ENV1_TIME_1"):getModulatorValue() is not needed.
When you attach the method to a modulator, the "value" argument contains the actual value of the modulator that changed
- your main "all env" modulator can have 0-99 values I suppose so you can simply attach a method like this to each all_env1_attack, all_env1_decay and so on.
Just include
```
Layer1Env1Attack:setValue(value, true)
Layer2Env1Attack:setValue(value, true)
Layer3Env1Attack:setValue(value, true)
...
Layer8Env1Attack:setValue(value, true)
```
- very simple but may need one method for each modulator in the All layer. So you can improve this further by using variables for your modulators or by setting them in custom groups

**With custom groups:**
- each modulator has a property "custom modulator index group"
- set all env 1 attack for layer 1 to 9 to some value say 1. decay to 2, sustain to 3 and so on
- create only one method that you can attach to the modulators in layer 9
- in that method, we will identify the group of the modulator we changed in layer 9 then apply the value of the "all" modulator to the layer 1-8 modulators
- by using that philosophy of custom groups, you can handle all you modulators in layer9 with only 1 method (nice ;-) )
- your method should look like this (I didn't test or tried it)

```
All_OnChange = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	-- No action if the panel is in bootstrap or program states (isPanelReady() is my own function)
	if not isPanelReady() then
		return
	end
	iGroup = mod:getPropertyInt("modulatorCustomIndexGroup")

	-- Loop on all modulators to change the values of those in the same group
	n = panel:getNumModulators()
		
	for i=0,n-1 do
		local modBI = panel:getModulatorByIndex(i)
		-- local Name= L(modBI:getName())
		if modBI:getPropertyInt("modulatorCustomIndexGroup") == iGroup then
			modBI:setValue(value, true)
		end
	end
end
```

Hope it helps...
/goodweather

**Comment by DRBAUS** [comment]
Hi goodweather and Tedjuh,

Thanks for the replies.

@Tedjuh1976 That method for val =0,127 works well.

@goodweather Yes, I have used the CustomIndexGroup in the Panel for the Mirage to display certain text and values in an LCD. 
I tried it out with the Panel for this unit and unfortunately it is really bad at handling sysex so I created a work-around which works.

I now have a working script in Layer 9 that sends out an os.execute(sleep(300)) after each layer's sysex message. It sends it directly to the sampler. In the script is also Tedjuh's method but instead of (value, true) with a (value, false) so that the Modulator values in Layer 1 - 8 get updated but don't fire their sysex scripts to the sampler. 

Thanks a lot. 

BAUS
